firefox-ios/Client/Frontend/Browser/Authenticator.swift

```diff
@@ -35,96 +40,120 @@ class Authenticator {
+                        return nil
                     }
-                    return nil
+                    loginsProvider.deleteLogins(ids: malformedGUIDs) { _ in }
                 }
-                loginsProvider.deleteLogins(ids: malformedGUIDs).upon { _ in }
-            }
 
-            // Found a single entry but the schemes don't match. This is a result of a schemeless entry that we
-            // saved in a previous iteration of the app so we need to migrate it. We only care about the
-            // the username/password so we can rewrite the scheme to be correct.
-            else if logins.count == 1 && logins[0].protectionSpace.`protocol` != challenge.protectionSpace.`protocol` {
-                let login = logins[0]
-                credentials = login.credentials
-                let new = LoginEntry(credentials: login.credentials, protectionSpace: challenge.protectionSpace)
-                return loginsProvider.updateLogin(id: login.id, login: new)
-                    >>> { deferMaybe(credentials) }
-            }
+                // Found a single entry but the schemes don't match. This is a result of a schemeless entry that we
+                // saved in a previous iteration of the app so we need to migrate it. We only care about the
+                // the username/password so we can rewrite the scheme to be correct.
+                else if logins.count == 1 && logins[0].protectionSpace.`protocol` != challenge.protectionSpace.`protocol` {
+                    let login = logins[0]
+                    credentials = login.credentials
+                    let new = LoginEntry(credentials: login.credentials, protectionSpace: challenge.protectionSpace)
+                    loginsProvider.updateLogin(id: login.id, login: new) { result in
+                        switch result {
+                        case .success:
+                            completionHandler(credentials)
+                        case .failure:
+                            completionHandler(nil)
+                        }
+                    }
+                    return
+                }
 
-            // Found a single entry that matches the scheme and host - good to go.
-            else if logins.count == 1 {
-                credentials = logins[0].credentials
-            } else {
-                logger.log("No logins found for Authenticator",
-                           level: .info,
-                           category: .webview)
-            }
+                // Found a single entry that matches the scheme and host - good to go.
+                else if logins.count == 1 {
+                    credentials = logins[0].credentials
+                } else {
+                    logger.log("No logins found for Authenticator",
+                               level: .info,
+                               category: .webview)
+                }
 
-            return deferMaybe(credentials)
+                completionHandler(credentials)
```

Comment:
Recommend adding a space below 153 just for a little bit more readability, it's a big block to parse when reading